les/peer.go

```diff
@@ -780,6 +794,34 @@ func (p *peer) Handshake(td *big.Int, head common.Hash, headNum uint64, genesis
 	return nil
 }
 
+// Returns true if the peer has indicated it is willing to transmit the given
+// transaction to the network. It may be the case that this client expects a
+// node to relay a transaction, but the server decides not to.
+func (p *peer) WillAcceptTransaction(tx *types.Transaction) bool {
+	if p.onlyAnnounce {
+		return false
+	}
+
+	// Retrieve the gateway fee information known for this peer.
+	gatewayFee := p.GatewayFee()
+	var etherbase common.Address
+	if ptr := p.Etherbase(); ptr != nil {
+		etherbase = *ptr
+	}
+
+	// Check that the transaction meets the peer's gateway fee requirements.
+	// Treat nil (i.e. unknown) peer gateway fee information as possibly free relay.
+	if etherbase != (common.Address{}) && gatewayFee != nil && gatewayFee.Cmp(common.Big0) > 0 {
```

Comment:
I remember there was a tricky interpretation on `etherbase`, something like:

```
etherbase == nil => not initialized yet (waiting from msg from the server)
etherbase == ZeroAddr => accept any tx
etherbase == someAddr => only accept tx with someAddr as feeRecipient
```

it's very subtle, and easy to miss... if possible i would make that clear now that we are doing more things with the etherbase.

Reply:
Here is a bit of a simplification 148acf3a75b44f241ce06aa157c779a1821612a8. In that commit I've changes it such that the `Etherbase` and `GatewayFee` methods return an `ok` boolean to indicate whether the values are known.